### ISSUE TITLE
4.25?

### ISSUE THREAD

**Issue by fllyingburritos** [UnrealEngineHTML5/Documentation#27]
Is there a way to get html5 in unreal engine 4.25?  It came out 2 days ago.

**Comment by eeench** [comment]
Yes I am also interested in this!! I cannot access the github html5 directory

**Comment by Simonkfc** [comment]
ill add myself to the line looking for this, so sad they removed html support

**Comment by JulienLussiez** [comment]
Same here.

**Comment by Mosel3y** [comment]
See here: https://github.com/UnrealEngineHTML5/Documentation/blob/master/Platforms/HTML5/HowTo/README.4.faq.UE4.HTML5.md#warning-this-is-not-for-the-faint-of-heart

ES2 is removed from 4.25. So getting HTML5 working would be a lot of work. It's not clear whether Nick is still working on this.

Unfortunately this isn't going to happen without developers working on it.

**Comment by fllyingburritos** [comment]
That's stupid, Epic Games said new vertions would be on github...

**Comment by fllyingburritos** [comment]
also what is ES2?  Sorry, I have no idea

**Comment by eeench** [comment]
Does anyone know the reason for this? I am now basically stuck with 4.23
for my WebGL projects...

**Comment by JulienLussiez** [comment]
This is not stupid but unfortunate. Epic didn't say that newer version would be on GitHub but that was now a community effort to make newer versions. I thought more developers would have worked on this but it seems that nobody cares so yes, we are stuck with 4.23. The fact that nobody cares is probably the reason why Epic dropped it.

**Comment by fllyingburritos** [comment]
4.24 works

**Comment by fllyingburritos** [comment]
there are definitely a lot of people that are interested in it though.   This video: https://www.youtube.com/watch?v=czirUc7tJpU&t=28s shows how to set up 4.24, and it has over 600 views.

**Comment by JulienLussiez** [comment]
Yes, I'm also very interested! Sometimes I participate in game jams and it's so much better to have a web version for this kind of event. Unfortunately, I had an issue with the 4.24 version but yes, I know it works but like Mosel3y said : "ES2 is removed from 4.25. So getting HTML5 working would be a lot of work." so without this work, we won't have a new version. I don't know if people are working on it. I hope so.

**Comment by fllyingburritos** [comment]
What is ES2?

**Comment by Mosel3y** [comment]
> What is ES2?

@fllyingburritos It refers to the OpenGL version.

https://en.wikipedia.org/wiki/OpenGL_ES

Unreal now uses ES3.1

**Comment by nickshinpho** [comment]
the **webgl** team (as well as the **emscripten** team) at **google** (and **mozilla** before them) was helping out with **epic** during and up to the **UE4.23** release.  there was a conscience decision to provide support for the "bare minimum" support "most" web browsers would support.  this was:

- **webgl (1) -> ES2**

(i can't remember when, but by) **UE4.24** HTML5 does however, defaults to:

- **webgl2 -> ES3**
  - (and i believe i removed the webgl 1 fallback code in 4.24 -- all in anticipation of getting ready to move to ES3.1)



anyways... the other major factor of why UE4's HTML5 plaform was deprecated was due to commercial/enterprise level usage (or lack thereof).  epic was spending a gigantic amount of testing resources -- 3 OS, 2 major browsers, with 2 additional (herding cats) ones -- for what seemed like little return (on projects targeting HTML5).

so it was a business decision to sunset it.


maybe when webGPU really gets going ... they will all bring it back alive ... (that's a big maybe...)

- if there's only a handful of people using it (versus millions like consoles and PCs... there's little reason for them to do so...)
- hence this fork to keep this going... just in case...


that said, today at the studio, i am busier than ever and cannot devote much time to UE4's HTML5 platform (whereas before, epic was bank rolling this effort).

- please be patient with me if i seem to drop off  -- i need to stay on top of the things that pay the billls...
- and i am appreciating more and more people that are able to help when they can


as for where "this" HTML5 is at for 4.25 -- i've started to see what it would take to get this up and running.

- i got it to the point of no compile errors -- but linking is a problem
- when i get to a successful link -- i'll push it up here
- to which i will need help from the public to get any run time issues figured out

thanks in advanced!

**Comment by Mosel3y** [comment]
Hey Nick, would like to say I appreciate your work on this. I understand you won't have as much time to do so. I'll be happy to help with issues and testing where possible.

**Comment by nickshinpho** [comment]
it is i who thanks you, Mosel3y! 

you have been very helpful to a number of users here.  ^_^

**Comment by nickshinpho** [comment]
progress update:

> i got it to the point of no compile errors -- but linking is a problem
> when i get to a successful link -- i'll push it up here
> to which i will need help from the public to get any run time issues figured out

i've got the link issue figured out a few weeks ago, and then proceeded to fight with the cooker.  (remember, ES2 is no longer a valid target -- i do not have the resources to keep "ES2 in" while **moving everything to ES3.1**).

finally, got the cooker squared away (as well as missing localization files) a couple of weeks ago.  i've ripped out everything ES2 and got HTML5 ready to use ES3.1 (i.e. webgl 2.0 compute  / webgpu?).  of course loading the game in the browser crashes.

figuring that i should then upgrade the emscripten toolchain before trying to apply the `web 2.0 compute` work-in-progress PR -- that's when i found out that **upstream** (clang11) is now the default compiler -- i.e. **fastcomp** (clang6?) was deprecated.

so i had to ensure when upgrading the toolchain (especially switch compiler backend -- upstream vs fastcomp), that it is done with a "working version" of UE4 -- which meant going back to 4.24.3's branch and try upgrading emscripten from there.

this was finally pushed up:

- https://github.com/UnrealEngineHTML5/UnrealEngine/commit/4dc8940793d62302f6ef4ca8b5325ae49e028641
  - UE4 (4.24.3)
  - emscripten 1.39.18
  - this should also show you what it takes when upgrading the emscirpten toolchain every now and then...
    - WARNING: do not use 1.39.17 -- it fails on a preprocessing issue during link time... (glad it was fixed in .18)

i will now be able to get back to the 4.25 branch -- and hopefully just apply the "webgl 2.0 compute" PR:

- https://github.com/emscripten-core/emscripten/pull/7612


but, i have my day job i need to get back to...  stay tuned!

**Comment by corneel** [comment]
I have been struggling with the issue for a while but today I was able to run the whole script. The two changes I made was to upgrade to the latest version of git (and of course git-bash) and to attempt the install on the C drive (also an SSD). 

I just leave this here for in case it helps someone else.

**Comment by fllyingburritos** [comment]
the beta for 4.26 just came out.

**Comment by corneel** [comment]
> progress update:
> 
> > i got it to the point of no compile errors -- but linking is a problem
> > when i get to a successful link -- i'll push it up here
> > to which i will need help from the public to get any run time issues figured out
> 
> i've got the link issue figured out a few weeks ago, and then proceeded to fight with the cooker. (remember, ES2 is no longer a valid target -- i do not have the resources to keep "ES2 in" while **moving everything to ES3.1**).
> 
> finally, got the cooker squared away (as well as missing localization files) a couple of weeks ago. i've ripped out everything ES2 and got HTML5 ready to use ES3.1 (i.e. webgl 2.0 compute / webgpu?). of course loading the game in the browser crashes.
> 
> figuring that i should then upgrade the emscripten toolchain before trying to apply the `web 2.0 compute` work-in-progress PR -- that's when i found out that **upstream** (clang11) is now the default compiler -- i.e. **fastcomp** (clang6?) was deprecated.
> 
> so i had to ensure when upgrading the toolchain (especially switch compiler backend -- upstream vs fastcomp), that it is done with a "working version" of UE4 -- which meant going back to 4.24.3's branch and try upgrading emscripten from there.
> 
> this was finally pushed up:
> 
> * [UnrealEngineHTML5/UnrealEngine@4dc8940](https://github.com/UnrealEngineHTML5/UnrealEngine/commit/4dc8940793d62302f6ef4ca8b5325ae49e028641)
>   
>   * UE4 (4.24.3)
>   * emscripten 1.39.18
>   * this should also show you what it takes when upgrading the emscirpten toolchain every now and then...
>     
>     * WARNING: do not use 1.39.17 -- it fails on a preprocessing issue during link time... (glad it was fixed in .18)
> 
> i will now be able to get back to the 4.25 branch -- and hopefully just apply the "webgl 2.0 compute" PR:
> 
> * [emscripten-core/emscripten#7612](https://github.com/emscripten-core/emscripten/pull/7612)
> 
> but, i have my day job i need to get back to... stay tuned!

Thank you so much for keeping this going. I am very keen on keeping this going. I have a large project that I want to put on the web that will be a good test and potentially help to make the case for having HTML5 / [some other tech] available for the browser.

**Comment by pingouin84** [comment]
hello, I am interested in your work and I wanted to know where you are on version 4.25? thank you in advance.

**Comment by beuman** [comment]
First, let me thanks @nickshinpho for what you are achieving here. I love the 4.25 version for its amazing integration with JetBrain's Rider IDE.

like @corneel I'm also working on a project that I want to see on both mobile AND desktop web browser. I think in the future more and more people will look into this. There are a lot of interesting use cases here. If I manage to get the funding I need for my project count on me to support this

**Comment by SeanDavidS** [comment]
Hello everyone, this HTML5 thing is so good, can't believe they removed it out of official support.
Is there any news on the 4.25 support for HTML5? Can't seem to find that anywhere

**Comment by fllyingburritos** [comment]
Now that 4.26 is out, is there work on that?

**Comment by corneel** [comment]
I am available to work on this again. @nickshinpho let me know how we can help. I believe WebAssembly is about to explode and it would be great if we can get this technology working in Unreal again.

**Comment by Jimbojib** [comment]
Could I port a 4.26 regular unreal to the 4.24 html5  unreal?

**Comment by Dmitry210** [comment]
@nickshinpho Hi! I want to thank you for your work! Can I find out about the current status?

**Comment by corneel** [comment]
Yes, I am also keen to get started again. Getting this to work on 5.0 would be absolutely awesome.

**Comment by mattlaverick** [comment]
I'm also interested in getting this working in 4.26 or knowing any alternatives. Will keep checking back here for any updates

**Comment by POTKIRLAND** [comment]
omg thank anyone who is working on this you are amazing people thank you thank you thank you!!

**Comment by muhammadammar4342** [comment]
Is there any updates for Unreal Engine 4.26?

**Comment by gurankas** [comment]
We have a project for 4.27 and were interested in making a html5. Are you still working to make it happen in the newer engine version after 4.24?

**Comment by sebalvarezSKS** [comment]
Hello,
I thought that the HTML5 engine in UE 4.24 was the same as 4.23, but reading this thread it seems to me that some improvements were made. Was one of them the ability to play video? That is broken as of 4.23. I have a video I put in a widget to play full screen and I have tried everything without success. The video plays fine in the preview, and macOS and Windows builds, but in HTML5, it only shows a plain white screen, whether it's full screen or in the game 3D space.
